### PR TITLE
d/aws_launch_template - filter support

### LINKED ISSUE #8367: data source aws_launch_template could be filtered on tags and create-time
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

I'd like to filter my launch templates using create-time and tags and that seems supported by the API

https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-launch-templates.html

The name would have to switched to optional so we can find stuff by tag.

### New or Affected Resource(s)

* `aws_launch_template`

### Potential Terraform Configuration

```hcl
data "aws_launch_template" "default" {
  tags {
    "ami"           = "ami-12983912839"
    "instance-type" = "r5"
  }
}
```

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #8367, Relates #11994

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
data_source_aws_launch_template - add `filter` support
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSLaunchTemplateDataSource_'
--- PASS: TestAccAWSLaunchTemplateDataSource_basic (47.77s)
--- PASS: TestAccAWSLaunchTemplateDataSource_filter_tags (48.57s)
--- PASS: TestAccAWSLaunchTemplateDataSource_filter_basic (47.98s)
```

### CODE REVIEW

**Comment by ewbankkit** [comment]
Verified acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSLaunchTemplateDataSource_' 
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccAWSLaunchTemplateDataSource_ -timeout 120m
=== RUN   TestAccAWSLaunchTemplateDataSource_basic
=== PAUSE TestAccAWSLaunchTemplateDataSource_basic
=== RUN   TestAccAWSLaunchTemplateDataSource_filter_basic
=== PAUSE TestAccAWSLaunchTemplateDataSource_filter_basic
=== RUN   TestAccAWSLaunchTemplateDataSource_filter_tags
=== PAUSE TestAccAWSLaunchTemplateDataSource_filter_tags
=== CONT  TestAccAWSLaunchTemplateDataSource_basic
=== CONT  TestAccAWSLaunchTemplateDataSource_filter_tags
=== CONT  TestAccAWSLaunchTemplateDataSource_filter_basic
--- PASS: TestAccAWSLaunchTemplateDataSource_basic (28.36s)
--- PASS: TestAccAWSLaunchTemplateDataSource_filter_tags (28.70s)
--- PASS: TestAccAWSLaunchTemplateDataSource_filter_basic (28.76s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	28.789s
```

File: aws/data_source_aws_launch_template.go [hunk 1]
```diff
@@ -336,19 +336,35 @@ func dataSourceAwsLaunchTemplate() *schema.Resource {
...
 
-	log.Printf("[DEBUG] Reading launch template %s", d.Get("name"))
+	filters, filtersOk := d.GetOk("filter")
+	name, nameOk := d.GetOk("name")
+	tags, tagsOk := d.GetOk("tags")
 
-	dlt, err := conn.DescribeLaunchTemplates(&ec2.DescribeLaunchTemplatesInput{
-		LaunchTemplateNames: []*string{aws.String(d.Get("name").(string))},
-	})
+	if !filtersOk && !nameOk && !tagsOk {
```
Comment (ewbankkit) [inline aws/data_source_aws_launch_template.go]: Can this now be done at `terraform plan` time with [`Schema.AtLeastOneOf`](https://github.com/hashicorp/terraform-plugin-sdk/blob/ac2f8d01136a1a0198e27d49ea0ea6ed986ea4ee/helper/schema/schema.go#L215-L228)?

Comment (ewbankkit) [reply aws/data_source_aws_launch_template.go]: See [this comment](https://github.com/terraform-providers/terraform-provider-aws/pull/12404#discussion_r393003198).

Comment (DrFaust92) [reply aws/data_source_aws_launch_template.go]: removed check entirely

**Comment by ewbankkit** [comment]
Re-validated after latest changes:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSLaunchTemplateDataSource_' 
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccAWSLaunchTemplateDataSource_ -timeout 120m
=== RUN   TestAccAWSLaunchTemplateDataSource_basic
=== PAUSE TestAccAWSLaunchTemplateDataSource_basic
=== RUN   TestAccAWSLaunchTemplateDataSource_filter_basic
=== PAUSE TestAccAWSLaunchTemplateDataSource_filter_basic
=== RUN   TestAccAWSLaunchTemplateDataSource_filter_tags
=== PAUSE TestAccAWSLaunchTemplateDataSource_filter_tags
=== CONT  TestAccAWSLaunchTemplateDataSource_basic
=== CONT  TestAccAWSLaunchTemplateDataSource_filter_basic
=== CONT  TestAccAWSLaunchTemplateDataSource_filter_tags
--- PASS: TestAccAWSLaunchTemplateDataSource_filter_basic (28.59s)
--- PASS: TestAccAWSLaunchTemplateDataSource_basic (28.78s)
--- PASS: TestAccAWSLaunchTemplateDataSource_filter_tags (28.99s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	29.041s
```

File: website/docs/d/launch_template.html.markdown [hunk 2]
```diff
@@ -18,11 +18,25 @@ data "aws_launch_template" "default" {
...
+
 ## Argument Reference
 
 The following arguments are supported:
 
-* `name` - (Required) The name of the launch template.
+* `name` - (Optional) The name of the launch template.
+* `filter` - (Optional) One or more name/value pairs to use as filters. There are
+several valid keys, for a full reference, check out
+[describe-launch-templates in the AWS CLI reference][1].
```
Comment (bflad) [inline website/docs/d/launch_template.html.markdown]: Can you please update this documentation to the longer form configuration block style? e.g. https://github.com/terraform-providers/terraform-provider-aws/pull/12400/files#diff-259bbf21af9ff942736c706b725f4d6dR76-R88 Thank you!

File: website/docs/d/launch_template.html.markdown [hunk 3]
```diff
@@ -18,11 +18,30 @@ data "aws_launch_template" "default" {
...
+}
+```
+
 ## Argument Reference
 
 The following arguments are supported:
 
-* `name` - (Required) The name of the launch template.
+* `name` - (Optional) The name of the launch template.
+* `filter` - (Optional) Configuration block(s) for filtering. Detailed below.
```
Comment (bflad) [inline website/docs/d/launch_template.html.markdown]: Oh ha, missing `tags` argument here -- will fix on merge. 👍 

```suggestion
* `filter` - (Optional) Configuration block(s) for filtering. Detailed below.
* `name` - (Optional) The name of the launch template.
* `tags` - (Optional) A mapping of tags, each pair of which must exactly match a pair on the desired Launch Template.
```

Comment (DrFaust92) [reply website/docs/d/launch_template.html.markdown]: i think i may have missed this on a bunch of them after the last fix. will check

**Pull request merged**